Grid-based dimension handling

mlange05 opened the pull request:
Requires #378.

The primary intent of this PR is to remove all global instances of `Dimension` objects and manage dedicated default dimensions (`x, y, z, t, time`) either via the `Grid` or the `Function` object that uses them. The most useful effect of this is that the `Operator`-level substitutions of spacing variable (`h_x, h_y, h_z`) now become optional, as the `Grid` can provide default values for them. Explicit spacing substitution is kept for seismic examples though, but provided via much neater API.

The changes in detail:
* The `Grid` now creates default space and time dimensions and all derived `Function` objects use those, unless explicitly overridden by the user.
* The spacing symbols on each `Dimension` are now created as `Constant` objects, allowing us to use the defaulting mechanism in the argument engine to provide `Grid`-dependent defaults.
  * As a corollary of the above, we can now **create completely state-free generic operators**, that have no hard-coded assumptions about the domain geometry in them. This is important for baking static libraries from Devito operators.
* The `Grid` now also provides a `spacing_map` that can be used to provide the operator-level substitutions of spacing symbols. This is used throughout the seismic examples.
  * This map does not include the temporal spacing, which has been renamed `s`->`dt`, so as to allow the notation of `op(time=ntimesteps, dt=dt)` on operator invocation.
* `SparseFunction` objects always need to be created from a `Grid`, since they derive all meta-information about their coordinate space from it (eg. problem dimension). This will likely have some overlap with #381.
* **No more global dimension objects!**
* Various small tweaks and fixes across the DSE/DLE and arguments engine to remove hard-coded use of default dimensions and/or fix resulting bugs.

FabioLuporini commented on tests/test_dle.py:
```diff
@@ -39,9 +39,8 @@ def simple_function(a, b, c, d, exprs, iters):
-    f = ResolveIterationVariable().visit(f, subs=subs)
-    f = SubstituteExpression(subs=subs).visit(f)
+    f = ResolveIterationVariable().visit(f)
+    f = SubstituteExpression().visit(f)
```
really? no ``subs``? how has this changed?

mlange05 replied:
Ooops, got caught in a blanket replace. Funny enough though, the tests still pass.... Anyway will put this back.

FabioLuporini commented on tests/test_dse.py:
```diff
@@ -164,8 +164,8 @@ def test_tti_rewrite_aggressive(tti_nodse):
-    ('shifted', 8, 364), ('shifted', 16, 830),
-    ('centered', 8, 170), ('centered', 16, 306)
+    ('shifted', 8, 355), ('shifted', 16, 811),
+    ('centered', 8, 168), ('centered', 16, 300)
```
better flop counts? 😮

mlange05 replied:
Yes, so the new method of substituting spacing symbols in the seismic stencils only captures space dimensions, not time as before. So `dt` is now a runtime argument, but there are only very few explicit multiplications with `dt` in the stencil, so they should not affect performance significantly.

Why it's going down instead of up, I don't know. Maybe the DSE finds better factorization/lifting opportunities if we don't pre-resolve `dt`?  Not sure though...

FabioLuporini replied:
or perhaps the baseline is lower

mlange05 commented on devito/dimension.py:
```diff
@@ -85,41 +113,33 @@ class LoweredDimension(Dimension):
-y = SpaceDimension('y', spacing=Symbol('h_y'))
-z = SpaceDimension('z', spacing=Symbol('h_z'))
 
 d = Dimension('d')
```
Yep, will fix.

mlange05 commented on devito/dse/graph.py:
```diff
@@ -228,14 +212,15 @@ def time_invariant(self, expr=None):
             return all(self.time_invariant(v) for v in self.values())
 
-        if any(i in expr.free_symbols for i in self.time_indices):
+        if any(isinstance(i, Dimension) and i.is_Time for i in expr.free_symbols):
```
Yes, that is the plan. I tried as part of #378, but ran out of patience. Will try again as a separate PR.

mlange05 commented on devito/dse/inspection.py:
```diff
@@ -101,7 +101,8 @@ def access(symbol):
 
     if mode in ['ideal', 'ideal_with_stores']:
-        filter = lambda s: t in s.atoms()
+        filter = lambda s: any(isinstance(Dimension, i) and i.is_Time
```
Ok, I'll see what I can do.

mlange05 commented on devito/grid.py:
```diff
@@ -48,8 +53,10 @@ class Grid(object):
+
     def __init__(self, shape, extent=None, origin=None, dimensions=None,
-                 dtype=np.float32):
+                 time_dimension=None, dtype=np.float32):
```
True, will do.

mlange05 commented on devito/grid.py:
```diff
@@ -58,8 +65,25 @@ def __init__(self, shape, extent=None, origin=None, dimensions=None,
+
+        # Store or create default symbols for time and stepping dimensions
+        if time_dimension is None:
+            self.time_dim = TimeDimension('time', spacing=Constant(name='dt'))
```
Yeah, a free function sounds sensible. The globally available dummy grid, however, won;t work, since the grid now adds default spacing values to the `Constant` spacing symbols, so it's not generic enough.

mlange05 commented on devito/operator.py:
```diff
@@ -83,6 +80,11 @@ def __init__(self, expressions, **kwargs):
         stencils = self._retrieve_stencils(expressions)
 
+        # Set the direction of time acoording to the given TimeAxis
+        for time in [d for d in self.dimensions if d.is_Time]:
```
Ok

FabioLuporini replied:
so, if you won't do it (and I accept it), should we file a reminder issue? because the more I look at this, the more it looks weird, if not bad

FabioLuporini commented:
Oh, don't forget to drop the spurious ``is_Fixed``, if you haven't done that already

mlange05 commented:
> Oh, don't forget to drop the spurious is_Fixed, if you haven't done that already

Yes, will do.

As for not using keyword args for `Function` constructors I would generally agree in principle, but I've tried before and it's a bit fiddly so I gave up. This kind of also ties in with the idea of not actually having a second-level cache, but using `_hashable_content` and maybe some `uuid` magic to force sympy to provide our desired caching behaviour for `Function` types. Both are on my wishlist, but quite complicated, and I have given up before. Maybe I should file an issue reminder...

mlange05 commented:
Ok, I believe I have addressed all comments. @FabioLuporini can you confirm please?

The final problem now is the travis YASK builder, which times out. The problem seems to be that the YASK backend can currently not deal with `Constant` symbols, thus it reverts to explicit point-wise `get/set` methods for the core stencil instead of `run_solution()`. This ultimately is super slow and causes timeouts.

FabioLuporini commented on devito/dse/graph.py:
```diff
@@ -228,14 +212,15 @@ def time_invariant(self, expr=None):
             temporaries = []
             for i in retrieve_terminals(item):
-                if any(j in i.free_symbols for j in self.time_indices):
+                if any(isinstance(j, Dimension) and j.is_Time for j in i.free_symbols):
```
you can use ``q_timedimension`` here

FabioLuporini commented:
Could you also drop ``is_Open`` and ``is_Closed`` in ``Iteration``? They're now useless

mlange05 commented:
Update: I believe we have found all remaining YASK errors due to `Constant` handling. This means we need #384 to go in first. I also had to disallow explicit `shape`/`dimension` parameters in the `TimeFunction` constructor, but that was pretty much unused anyway. Let's see if travis agrees with my assessment.

FabioLuporini commented:
Yes, please rebase and then, if Travis permits, merge.

mlange05 commented:
Ok, let's go. Merging...

mloubout commented:
>This means we need #384 to go in first. I also had to disallow explicit shape/dimension parameters in the TimeFunction constructor, but that was pretty much unused anyway. 

Since when, these PR forcing are getting really annoying,

mlange05 commented:
@mloubout The requirement is simply due to a YASK-bug that was fixed in #384 and was causing travis to fail for this branch. Creating and merging a separate branch is actually good practice, since it allows a single feature or fix to be reviewed independently, as in this case the use of `Constant` symbols through YASK. If another branch requires a certain feature or fix, this simply creates a dependency, and thus forces the merge order. This is the essence of distributed development and a vastly superior workflow to creating huge monolithic branches/PRs that contain multiple features and fixes.

mloubout commented:
>The requirement is simply due to a YASK-bug that was fixed in #384 

If it was fixed , why is the requirement still there?

mlange05 commented:
??? #384 fixes the bug, so I rebased this branch on top of #384. Merging the individual branches in order is simply good practice to keep the diff clean and and managable. I really don't understand what your problem is here?

mloubout commented:
>order is simply good practice to keep the diff clean and and managable

I have no issue with that at all, my question is:

>I also had to disallow explicit shape/dimension parameters in the TimeFunction constructor

is this no decided as not possible?

mlange05 commented:
Ah, now I get what you mean now. Yes, nothing in the test base really required explicit `shape`/`dimension` pairs for `TimeFunction`, so I removed it to break the cyclic dependency between `Function` and `Constant` that arose from the defaulting code for time (both need to be dynamically dispatched through `base.py` for the backend to work correctly).

If you need to define explicit dimensions on a `TimeFunction` object can you please provide an example and/or test case so that we can find another solution for this? In the meantime, it should need a minimal coding overhead to force `TimeFunction` behaviour with custom dimensions via the `Function` interface, for example (untested):
```
grid = Grid(....)
x, y, z = grid.dimensions
t = grid.stepping_dim
f = <some custom dimension>
u = Function(name='u', dimensions=(t, f, x, y, z), ...)
```

As for the "forcing" this change, I fast-tracked this PR because the infamous #345 now depends on it. Sorry, if that decision was rushed and/or if this inconveniences you. But I believe we all want want #345 out of the way now... :frowning:

mloubout commented:
Great, thanks for the detailed answer, will cook up a test/example as soon as I find time
